### PR TITLE
Add Goal Count Heuristic

### PR DESCRIPTION
**Pull request by williamshen-nz**
Duplicate of https://github.com/Learning-and-Intelligent-Systems/predicators/pull/494 but on a branch on the main repository

### CODE REVIEW

**Review by ronuchit — APPROVED** [review]
LGTM! If you want the full experience, you can try a supercloud run (see `analysis/run_supercloud_experiments.sh`). Would be fun to see the results.

File: src/utils.py [hunk 1]
```diff
@@ -1065,6 +1067,19 @@ def __call__(self, atoms: Collection[GroundAtom]) -> float:
         raise NotImplementedError("Override me!")
 
 
+class GoalCountHeuristic(_TaskPlanningHeuristic):
+    """The number of goal atoms that are not in the current state."""
+    HEURISTIC_NAME: str = "goal_count"
+
+    @functools.cached_property
+    def _goal_set(self) -> Set[GroundAtom]:
+        """Convert the goal atom Collection into an explicit Set."""
```
Comment (tomsilver) [inline src/utils.py:1076]: goals should already be `Set[GroundAtom]` (see the definition of `Task` in `structs.py` for example). did you run into a type checking issue without this? if so, we should probably change whichever Collection is causing the issue.

File: tests/test_utils.py [hunk 2]
```diff
@@ -1339,37 +1339,86 @@ def test_operator_application():
     assert next_atoms == {pred1([cup1, plate1]), pred2([cup1, plate1])}
 
 
-def test_create_task_planning_heuristic():
+@pytest.mark.parametrize("heuristic_name, expected_heuristic_cls", [
```
Comment (tomsilver) [inline tests/test_utils.py:1342]: this is awesome, thank you for introducing the pattern!

File: tests/test_utils.py [hunk 3]
```diff
@@ -1339,37 +1339,86 @@ def test_operator_application():
...
+        GroundAtom(on, [block3, table2]),
+    }) == 2
+    assert heuristic({
+        GroundAtom(on, [block1, table2]),
+        GroundAtom(on, [block2, table2]),
+        GroundAtom(on, [block3, table1]),
+    }) == 3
+
+    # Some edge cases
+    assert heuristic({}) == 3
```
Comment (tomsilver) [inline tests/test_utils.py:1416]: should this be `set()` instead?

**Review by tomsilver — COMMENTED** [review]
awesome work! a few questions

File: src/utils.py (hunk 1 above)
Comment (williamshen-nz) [reply src/utils.py:1076]:
```
@dataclass(frozen=True)
class _TaskPlanningHeuristic:
    """A task planning heuristic."""
    name: str
    init_atoms: Collection[GroundAtom]
    goal: Collection[GroundAtom]
    ...
```

It's still a `Collection[GroundAtom]` in the parent class which is why I added this here. That being said, `goal: Set[GroundAtom]` in `create_task_planning_heuristic` so we could change the typing to be a `Set` rather than a `Collection`

File: tests/test_utils.py (hunk 3 above)
Comment (williamshen-nz) [reply tests/test_utils.py:1416]: Yes! Good catch

File: src/utils.py (hunk 1 above)
Comment (tomsilver) [reply src/utils.py:1076]: ah I see, yeah that was our mistake. would you mind changing it there to be a Set? thanks!

Comment (ronuchit) [reply src/utils.py:1076]: yeah, there are similarly two instances of `git grep "init_atoms: Colle"`, we should fix both as well!

File: src/utils.py [hunk 4]
```diff
@@ -1037,6 +1037,10 @@ def ops_and_specs_to_dummy_nsrts(
     return nsrts
 
 
+# Note: create separate `heuristics.py` module if we need to add new
+#  heuristics in the future.
```
Comment (tomsilver) [inline src/utils.py:1041]: nit: extra space here

**Review by tomsilver — APPROVED** [review]
LGTM thanks again!

File: src/utils.py (hunk 4 above)
Comment (williamshen-nz) [reply src/utils.py:1041]: pycharm only highlights multiline notes when there's two spaces 😢
